### ISSUE TITLE
Remove the `MediaTrackSupportedConstraints` directionary

### ISSUE THREAD

**Issue by skyclouds2001** [mdn/content#30961]
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getSupportedConstraints

### What information was incorrect, unhelpful, or incomplete?

the `MediaTrackSupportedConstraints` is defined as a dictionary, and is only used by `MediaDevices.getSupportedConstraints()` as its return value

as https://github.com/mdn/content/discussions/5828 discusses, it is fine to remove this dictionary and plae its content in `MediaDevices.getSupportedConstraints()` *Return value* section

note that its compat data also need to remove

### What did you expect to see?

as above

**Comment by skyclouds2001** [comment]
`MediaTrackConstraints` dictionary is used in `MediaDevices.getUserMedia()` `MediaStreamTrack.getConstraints()` `MediaStreamTrack.applyConstraints()`, and is different from the spec defines, which should contain a `advanced` member and should inherit from `MediaTrackConstraintSet` dictionary

`MediaTrackSettings` is only used in `MediaStreamTrack.getSettings()`

**Comment by skyclouds2001** [comment]
`MediaTrackCapabilities` already removed, is used in `InputDeviceInfo.getCapabilities()` `MediaStreamTrack.getCapabilities()`

**Comment by hamishwillee** [comment]
Yes, at high level this makes sense, and BCD is already heading this way. 

The devil is in the detail - but I think most of that [problematic detail is in BCD](https://github.com/mdn/browser-compat-data/pull/21761#issuecomment-2043920678)

**Comment by Josh-Cena** [comment]
@hamishwillee @bsmth I'm seeking to fix this issue, because currently our architecture is quite messy. Do you think we should keep the subarticles for each dictionary member, or if I should just fold everything into the return value/parameters sections of the respective methods?

**Comment by chrisdavidmills** [comment]
I also support folding the properties back into the return value/parameter section of each relevant method. It is repetitive, but this is a tricky case, where some of the items aren't supported as constraints, capabilities, _and_ settings.

It will also make the BCD situation more manageable. I fell foul of this recently; see https://github.com/mdn/browser-compat-data/pull/27092.

**Comment by Josh-Cena** [comment]
The traditional approach is to find somewhere canonical that accepts the dictionary and put it there. Unfortunately for example, we have the same info duplicated under https://developer.mozilla.org/en-US/docs/Web/API/Request/Request and https://developer.mozilla.org/en-US/docs/Web/API/Window/fetch BCD tables, and so was content before we decided to bring back the `RequestInit` dictionary—if I were to do it today, I would write it once under `Request()` (same with BCD), and just make `fetch()` have no subproperties. For example, here `MediaTrackSupportedConstraints` and `MediaTrackSettings` are trivial, and we could decide to put all `MediaTrackConstraints` stuff under `applyConstraints` (which we already do).

**Comment by chrisdavidmills** [comment]
> The traditional approach is to find somewhere canonical that accepts the dictionary and put it there. Unfortunately for example, we have the same info duplicated under https://developer.mozilla.org/en-US/docs/Web/API/Request/Request and https://developer.mozilla.org/en-US/docs/Web/API/Window/fetch BCD tables, and so was content before we decided to bring back the `RequestInit` dictionary—if I were to do it today, I would write it once under `Request()` (same with BCD), and just make `fetch()` have no subproperties. For example, here `MediaTrackSupportedConstraints` and `MediaTrackSettings` are trivial, and we could decide to put all `MediaTrackConstraints` stuff under `applyConstraints` (which we already do).

I totally agree with you, under most circumstances. However, this situation is trickier. `screenPixelRatio`, for example, can only be used as a setting, not a constraint or capability, so you'd need at least two canonical places.

**Comment by Josh-Cena** [comment]
No we would have three canonical places anyway: for supported constraints, settings, and constraints. I'm saying whether we should duplicate the constraints info under every single place that accepts or returns it (which we do for `RequestInit`).

**Comment by chrisdavidmills** [comment]
> No we would have three canonical places anyway: for supported constraints, settings, and constraints. I'm saying whether we should duplicate the constraints info under every single place that accepts or returns it (which we do for `RequestInit`).

OK, sure, makes sense.

Would you have three canonical places just in BCD, or in the docs as well?

**Comment by Josh-Cena** [comment]
That part I'm unsure about, but I'm leaning towards three and only three canonical places for content, because as you said there are always these nuances (plus the types are different). For BCD, because there's no way to "cross-link", they may decide to duplicate for every single usage site. I won't touch BCD here though, since they don't have dictionaries anyway.

**Comment by chrisdavidmills** [comment]
> That part I'm unsure about, but I'm leaning towards three and only three canonical places for content, because as you said there are always these nuances (plus the types are different). For BCD, because there's no way to "cross-link", they may decide to duplicate for every single usage site. I won't touch BCD here though, since they don't have dictionaries anyway.

I think that's reasonable. Figuring out three sensible places to place the content, which can then be linked to from other relevant places, will likely help to inform the BCD structure anyway.

**Comment by hamishwillee** [comment]
^^^ Works as long as they content really is always the same in all uses of the dictionary. This was part of the problem when we used to have the macro to copy content - we'd duplicate - but the actual implementations would be slightly different and we ended up with theoretically the same thing that actually wasn't. Same problem occurs if you have a theoretically canonical doc and link.

The decision when we went around this was that duplication should be done in all places the dictionary was used, with no cross links. Then later on we decided that dictionaries are OK if things become unmanageable - so perhaps the rules are not so hard.

Upshot though, I'm in favour of trying the suggestion out ^^^.
